### PR TITLE
Spec Runtime Metadata structure

### PR DESCRIPTION
**Pull request by lamafab**
As requested. Still needs a little bit of work but almost done.

Spec is based on the structure V14 in https://github.com/paritytech/frame-metadata

### CODE REVIEW

**Comment by lamafab** [comment]
FYI: rebased on newest changes to `main`

File: 02_runtime/10-metadata.adoc [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+where::
+ * stem:[M] are the first four constant bytes, spelling "meta" in ASCII
+ * stem:[v_m] is the format version of the metadata (one unsigned byte, currently `14`)
+
+ * stem:[R] is a sequence (<<defn-scale-list>>) of type definitions stem:[r_i] (<<defn-rtm-registry-entry>>).
+
+ * stem:[P] is a sequence (<<defn-scale-list>>) of pallet metadata stem:[p_i] (<<sect-rtm-pallet-metadata>>).
+
+ * stem:[t_e] is the type Id (<<defn-rtm-type-id>>) of the extrinsics.
+ * stem:[v_e] is the format version of the extrinsics
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc]: is v_e also an unsigned byte, like v_m (version of metadata)?

File: 02_runtime/10-metadata.adoc [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+
+[stem]
+++++
+r_i = (i_t,p,T,D,c) \
+T = (t_0, ...,t_n) \
+t_i = (n, y)
+++++
+
+where::
+* stem:[i_t] is a compact integer indicating the identifier of the type.
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc]: maybe [id_t] is more appropriate for identifier than [i_t]. does the subscript t here denote the type of r_i?

File: 02_runtime/10-metadata.adoc [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+* stem:[p] is the path of the type, optional and based on source file location. Encoded a sequence (<<defn-scale-list>>) of strings.
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc]: typo: * Encoded as a sequence....

File: 02_runtime/10-metadata.adoc [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+* stem:[T] is a sequence (<<defn-scale-list>>) of generic parameters (empty for non-generic types).
+** stem:[n] is the name string of the generic type parameter
+** stem:[y] is a _Option_ type containing a type Id (<<defn-rtm-type-id>>).
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:99]: at the first glance definition 165 and 166 seem to be cyclical. Is there a need for defining 166 separately? Isn't Def-166 same as the i_t defined in Def-165. If both are needed, consider if its more apt to have 166 defined before 165.

File: 02_runtime/10-metadata.adoc [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+
+.Type Variant
+[#defn-rtm-type-definition]
+====
+The type definition stem:[D] indicates all the possible types of encodable
+values a type can have.
+
+[stem]
+++++
+D = {
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:152]: Clarify if D is just a single unsigned byte?

File: 02_runtime/10-metadata.adoc [hunk 6]
```diff
@@ -0,0 +1,727 @@
...
+S = (l, y)
+++++
+** stem:[l] is a unsigned 32-bit integer indicating the length
+** stem:[y] is a type Id (<<defn-rtm-type-id>>).
+* stem:[T] is a sequence (<<defn-scale-list>>) of type Ids (<<defn-rtm-type-id>>).
+* stem:[P] is a varying datatype of the following structure:
++
+[stem]
+++++
+P = {
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:196]: I could not parse the definition of P. Does it make sense to have arrows between 0 and boolean, 1 and char, so on (like how id defined for D)?

File: 02_runtime/10-metadata.adoc [hunk 7]
```diff
@@ -0,0 +1,727 @@
...
++
+[stem]
+++++
+B = (s, o)
+++++
++
+** stem:[s] is a type Id (<<defn-rtm-type-id>>) representing the bit store
+type (https://docs.rs/bitvec/latest/bitvec/store/trait.BitStore.html[external
+reference])
+** stem:[o] is a type Id (<<defn-rtm-type-id>>) the bit order type
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:225]: typo: * representing the bit order ....

File: 02_runtime/10-metadata.adoc [hunk 8]
```diff
@@ -0,0 +1,727 @@
...
+The type definition stem:[D] indicates all the possible types of encodable
+values a type can have.
+
+[stem]
+++++
+D = {
+	(0,->,C,"composite type (e.g. structure or tuple)"),
+	(1,->,V,"variant type"),
+	(2,->,s_v,"sequence type varying length"),
+	(3,->,S,"sequence with fixed length"),
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:156]: Maybe nice to be consistent with the MetadataTypeDefinition diagram and add 'Array' in brackets next to D=3 -> Sequence with fixed length. In general make notations in Definition of D consistent with the diagram

File: 02_runtime/10-metadata.adoc [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+++++
+v_i = (n,F,k,c)
+++++
+
+where
+
+* stem:[n] is a string representing the name of the variant.
+* stem:[F] is a possible empty array of varying length containing field
+(<<defn-rtm-field>>) elements.
+* stem:[k] is an unsigned 8-bit integer indicating the index of the variant (TODO: Clarify).
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc]: Remove the TODO

File: 02_runtime/10-metadata.adoc [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:376]: Define the parameter c

File: 02_runtime/10-metadata.adoc [hunk 11]
```diff
@@ -0,0 +1,727 @@
...
+      storage_modifier:
+        0: optional
+        1: default
+++++
+====
+
+.Storage Entry Modifier
+[#defn-rtm-storage-entry-modifier]
+====
+
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:560]: This needs to be re-looked into. What are 0 and 1 the options for?

File: 02_runtime/10-metadata.adoc [hunk 12]
```diff
@@ -0,0 +1,727 @@
...
+where _0_ indicates that the entry returns an _Option_ type and therefore _None_
+if the storage entry is not present. _1_ indicates that the entry returns the
+type stem:[y] with default value stem:[D] (in
+<<defn-rtm-storage-entry-metadata>>) if the entry is not present.
+====
+
+.Storage Entry Type
+[#defn-rtm-storage-entry-type]
+====
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:583]: same issue as definition 172

File: 02_runtime/10-metadata.adoc [hunk 13]
```diff
@@ -0,0 +1,727 @@
...
+        enum: hasher_type
+        repeat: expr
+        repeat-expr: num_hasher.value
+
+      - id: key
+        type: scale::compact_int
+      - id: value
+        type: scale::compact_int
+    enums:
+      hasher_type:
```
Comment (bhargavbh) [inline 02_runtime/10-metadata.adoc:637]: same issue as Def-172

File: 02_runtime/10-metadata.adoc (hunk 2 above)
Comment (lamafab) [reply 02_runtime/10-metadata.adoc]: No. Updated.

File: 02_runtime/10-metadata.adoc (hunk 4 above)
Comment (lamafab) [reply 02_runtime/10-metadata.adoc:99]: Definition 166 is referenced all over the place, so definition 165 and 166 are unrelated.

> Isn't Def-166 same as the i_t defined in Def-165

No, Def 166 indicates the entry in the sequences (by index), while i_t refers to another entry. I expanded on that definition.

File: 02_runtime/10-metadata.adoc (hunk 6 above)
Comment (lamafab) [reply 02_runtime/10-metadata.adoc:196]: I'm not quite sure what you mean here? It's a varying datastructure indicating the primitive type (`0` means boolean, `1` means char, etc)

File: 02_runtime/10-metadata.adoc (hunk 8 above)
Comment (lamafab) [reply 02_runtime/10-metadata.adoc:156]: I guess so, but this should be updated in the Kaitai structure. We use the word "sequence" almost exclusively.

**Comment by FlorianFranzen** [comment]
Please also include anything mentioned in #539.

**Comment by lamafab** [comment]
Done and done. Besides the two open points I asked Bhargav, I think we're ready

File: 02_runtime/10-metadata.adoc (hunk 8 above)
Comment (FlorianFranzen) [reply 02_runtime/10-metadata.adoc:156]: Yes, that is my mistake. The kaitai struct definition was partial written of polkadot-js json data, so I might have messed up some of the names. Feel free to change the definition.

**Review by bhargavbh — APPROVED** [review]

**Comment by FlorianFranzen** [comment]
Rebased on latest main, ready to be merged.

**Review by FlorianFranzen — APPROVED** [review]

**Pull request merged**